builtin/logical/database/path_rotate_credentials.go

```diff
@@ -78,22 +78,30 @@ func (b *databaseBackend) pathRotateRootCredentialsUpdate() framework.OperationF
 			return nil, err
 		}
 
-		// Take out the backend lock since we are swapping out the connection
-		b.Lock()
-		defer b.Unlock()
-
 		// Take the write lock on the instance
 		dbi.Lock()
-		defer dbi.Unlock()
-
+		unlocked := false
+		defer func() {
+			if !unlocked {
+				// set to true before unlocking to avoid race condition
+				unlocked = true
+				dbi.Unlock()
+			}
```

Comment:
It is, because the other `defer` could panic, and we want to ensure we *always* unlock.

Reply:
The initial defer call (i.e. `dbi.Unlock()`) will get invoked since it's already pushed to the stack, so even if it panics during the second defer (e.g. by the `Close` call) it should safely `dbi.Unlock()` without us having to call it explicitly within that second defer func on L99.

Example: https://go.dev/play/p/-V8L4XnrFHD

Reply:
We need to unlock before we call `b.ClearConnectionId()`, but we can't guarantee that `dbi.database.Close()` won't panic, and we also can't guarantee that the other `defer` runs first.

I suppose we could combine them and move the `unlock()` call to the top of this `defer`, and just remove the other `defer` entirely?

Reply:
We chatted about this and sorted it out into simpler `defer`s.